pkg/agent/multicast/mcast_discovery.go

```diff
@@ -201,6 +214,42 @@ func (s *IGMPSnooper) collectStats() (igmpANPStats, igmpACNPStats map[apitypes.U
 	return igmpANPStats, igmpACNPStats
 }
 
+func (s *IGMPSnooper) sendIGMPReport(groupRecordType uint8, groups []net.IP) error {
+	igmp, err := s.generateIGMPReportPacket(groupRecordType, groups)
+	if err != nil {
+		return err
+	}
+	if err := s.ofClient.SendIGMPRemoteReportPacketOut(igmpReportDstMac, types.IGMPv3Router, igmp); err != nil {
+		return err
+	}
+	klog.V(2).InfoS("Sent packetOut for IGMP v3 report", "groups", groups)
+	return nil
+}
+
+func (s *IGMPSnooper) generateIGMPReportPacket(groupRecordType uint8, groups []net.IP) (util.Message, error) {
+	records := make([]protocol.IGMPv3GroupRecord, len(groups))
+	for i, group := range groups {
+		records[i] = protocol.IGMPv3GroupRecord{
+			Type:             groupRecordType,
+			MulticastAddress: group,
+		}
+	}
+	return &protocol.IGMPv3MembershipReport{
+		Type:           protocol.IGMPv3Report,
+		Checksum:       0,
+		NumberOfGroups: uint16(len(records)),
+		GroupRecords:   records,
+	}, nil
+}
+
+func (s *IGMPSnooper) sendIGMPJoinReport(groups []net.IP) error {
+	return s.sendIGMPReport(protocol.IGMPIsEx, groups)
```

Comment:
why it sends Exclude record with the provided groups when it joins them?

Reply:
I didn't set any sources in function `generateIGMPReportPacket`, which means the report message does not have any sources configured. 

The IGMPv3GroupRecord struct is like this,
```
type IGMPv3GroupRecord struct {
	Type             uint8
	AuxDataLen       uint8 // this should always be 0 as per IGMPv3 spec.
	NumberOfSources  uint16
	MulticastAddress net.IP
	SourceAddresses  []net.IP
	AuxData          []uint32 // NOT USED
}
```
`SourceAddresses` in the generated packet is nil according to my implementation.